### PR TITLE
fix(typescript): accept union types

### LINKED ISSUE #8: type definitions don't work for union types
```ts
arrify(Boolean() ? [1, 2] : 3);
```

The above returns the type `(number | number[])[]`. This doesn't match the runtime value of `number[]`

### PR DESCRIPTION
**Pull request by callmehiphop**
Fixes #8

Please be gentle!

### CODE REVIEW

**Comment by sindresorhus** [comment]
// @BendingBender

File: index.d.ts [hunk 1]
```diff
@@ -28,5 +28,6 @@ declare function arrify(value: string): [string];
 declare function arrify<ValueType>(value: ValueType[]): ValueType[];
 declare function arrify<ValueType>(value: Iterable<ValueType>): ValueType[];
 declare function arrify<ValueType>(value: ValueType): [ValueType];
+declare function arrify<ValueType>(value: ValueType|ValueType[]): ValueType[];
```
Comment (BendingBender) [inline index.d.ts:31]: I think that we should rethink the whole definion. What you propose only fixes the case that the parameter is of a special type or an array of this type (`number | number[]`). What about `number | string[]` or any other type combination?

I've come up with this one, it seems to work as expected but is very ugly:
```suggestion
declare function arrify<ValueType>(
	value: ValueType
): ValueType extends (null | undefined)
	? []
	: ValueType extends string
	? [string]
	: ValueType extends (infer T)[]
	? T[]
	: ValueType extends ReadonlyArray<infer T>
	? ReadonlyArray<T>
	: ValueType extends Iterable<infer T>
	? T[]
	: [ValueType];
```

Do you have a better idea?

File: index.test-d.ts [hunk 2]
```diff
@@ -12,3 +12,4 @@ expectType<([string | number, string | number])[]>(
 	arrify(new Map<string | number, string | number>([[1, 2], ['a', 'b']]))
 );
 expectType<number[]>(arrify(new Set([1, 2])));
+expectType<number[]>(arrify<number>(Boolean() ? [1, 2] : 3));
```
Comment (BendingBender) [inline index.test-d.ts:15]: Following tests pass with just the proposed type above, no other overloads needed:
```suggestion
expectType<[]>(arrify(null));
expectType<[]>(arrify(undefined));
expectType<[string]>(arrify('🦄'));
expectType<string[]>(arrify(['🦄']));
expectType<[boolean]>(arrify(true));
expectType<[number]>(arrify(1));
expectType<[{}]>(arrify({}));
expectType<([string | number, string | number])[]>(
	arrify(new Map<string | number, string | number>([[1, 2], ['a', 'b']]))
);
expectType<number[]>(arrify(new Set([1, 2])));
expectError(arrify(['🦄'] as const).push(''));
expectType<number[] | []>(arrify(Boolean() ? [1, 2] : null));
expectType<number[] | []>(arrify(Boolean() ? [1, 2] : undefined));
expectType<number[] | [string]>(arrify(Boolean() ? [1, 2] : '🦄'));
expectType<number[] | string[]>(arrify(Boolean() ? [1, 2] : ['🦄']));
expectType<number[] | [boolean]>(arrify(Boolean() ? [1, 2] : true));
expectType<number[] | [number]>(arrify(Boolean() ? [1, 2] : 3));
expectType<number[] | [{}]>(arrify(Boolean() ? [1, 2] : {}));
expectType<number[] | (string | boolean)[]>(
	arrify(Boolean() ? [1, 2] : new Set<string | boolean>(['🦄', true]))
);
expectType<number[] | [boolean] | [string]>(
	arrify(Boolean() ? [1, 2] : Boolean() ? true : '🦄')
);
```

**Review by BendingBender — CHANGES_REQUESTED** [review]

File: index.d.ts (hunk 1 above)
Comment (sindresorhus) [reply index.d.ts:31]: // @ajafff

Comment (ajafff) [reply index.d.ts:31]: > : ValueType extends Iterable<infer T>
> ? [T]

`[T]` should be `T[]`

Comment (BendingBender) [reply index.d.ts:31]: Right, thank you!

Comment (BendingBender) [reply index.d.ts:31]: @callmehiphop Could you please update your PR with the suggested changes?

Comment (callmehiphop) [reply index.d.ts:31]: I'm ok closing this one out if you want to open a new PR with those changes

Comment (BendingBender) [reply index.d.ts:31]: Ok, close this one, I'll open a new one.

**Pull request closed without merging**